src/workarounds/forwarddiff_rules.jl

```diff
@@ -191,54 +191,69 @@ function construct_value(basis::PlaneWaveBasis{T}) where {T <: ForwardDiff.Dual}
                    comm_kpts=basis.comm_kpts)
 end
 
+
 function self_consistent_field(basis_dual::PlaneWaveBasis{T};
                                response=ResponseOptions(),
                                kwargs...) where T <: ForwardDiff.Dual
     # Note: No guarantees on this interface yet.
 
     # Primal pass
-    basis  = construct_value(basis_dual)
-    scfres = self_consistent_field(basis; kwargs...)
-
-    ## promote occupied bands to dual numbers
-    occupation_dual = [T.(occₖ) for occₖ in scfres.occupation]
-    ψ_dual = [Complex.(T.(real(ψₖ)), T.(imag(ψₖ))) for ψₖ in scfres.ψ]
-    ρ_dual = DFTK.compute_density(basis_dual, ψ_dual, occupation_dual)
-    εF_dual = T(scfres.εF)  # Only needed for entropy term
-    eigenvalues_dual = [T.(εₖ) for εₖ in scfres.eigenvalues]
-    energies_dual, ham_dual = energy_hamiltonian(basis_dual, ψ_dual, occupation_dual;
-                                                 ρ=ρ_dual, eigenvalues=eigenvalues_dual,
-                                                 εF=εF_dual)
-
-    response.verbose && println("Solving response problem")
+    basis_primal  = construct_value(basis_dual)
+    scfres = self_consistent_field(basis_primal; kwargs...)
+
+    ## Compute external perturbing potential δV_{ext} (contained in δH)
+    ## and form the matrix-vector product with the bands
+    δH, δHψ = let
+        occupation_dual = [T.(occk) for occk in scfres.occupation]
+        ψ_dual = [Complex.(T.(real(ψk)), T.(imag(ψk))) for ψk in scfres.ψ]
+        ρ_dual = DFTK.compute_density(basis_dual, ψ_dual, occupation_dual)
+        εF_dual = T(scfres.εF)  # Only needed for entropy term
+        eigenvalues_dual = [T.(εk) for εk in scfres.eigenvalues]
+        _, δH = energy_hamiltonian(basis_dual, ψ_dual, occupation_dual;
```

Comment:
So all this is tricky because this is not really a dH by itself, right? It's a Dual(H, dH). So, in keeping with the rest, dH -> H_dual?